include/net/cloud.h

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#ifndef ZEPHYR_INCLUDE_CLOUD_H_
+#define ZEPHYR_INCLUDE_CLOUD_H_
+
+/* Cloud API */
+
+#include <zephyr.h>
+
+/**@brief Cloud backend states. */
+enum cloud_state {
+	CLOUD_STATE_DISCONNECTED,
+	CLOUD_STATE_DISCONNECTING,
+	CLOUD_STATE_CONNECTED,
+	CLOUD_STATE_CONNECTING,
+	CLOUD_STATE_BUSY,
+	CLOUD_STATE_ERROR,
+	CLOUD_STATE_COUNT
+};
+
+/**@brief Cloud events that can be notified asynchronously by the backend. */
+enum cloud_event_type {
+	CLOUD_EVT_CONNECTED,
+	CLOUD_EVT_DISCONNECTED,
+	CLOUD_EVT_READY,
+	CLOUD_EVT_ERROR,
+	CLOUD_EVT_DATA_SENT,
+	CLOUD_EVT_DATA_RECEIVED,
+	CLOUD_EVT_COUNT
+};
+
+/**@brief Quality of Service for message sent by a cloud backend. */
+enum cloud_qos {
+	CLOUD_QOS_AT_MOST_ONCE,
+	CLOUD_QOS_AT_LEAST_ONCE,
+	CLOUD_QOS_EXACTLY_ONCE,
+	CLOUD_QOS_COUNT
+};
+
+/**@brief Quality of Service for message sent by a cloud backend. */
+enum cloud_resource {
+	CLOUD_EP_TOPIC,
+	CLOUD_EP_URI,
+	CLOUD_EP_COUNT
+};
+
+/* Forward declaration of cloud backend type. */
+struct cloud_backend;
+
+/**@brief Cloud message type. */
+struct cloud_msg {
+	char *payload;
+	size_t len;
+	enum cloud_qos qos;
+	struct {
+		enum cloud_resource type;
+		char *str;
+		size_t len;
+	} resource;
+};
+
+/**@brief Cloud event type. */
+struct cloud_event {
+	enum cloud_event_type type;
+	union {
+		struct cloud_msg msg;
+		int err;
+	} data;
+};
+
+/**
+ * @brief Cloud event handler function type.
+ *
+ * @param[in] backend Pointer to cloud backend.
+ * @param[in] evt Pointer to cloud event.
+ */
+typedef void (*cloud_evt_handler_t)(const struct cloud_backend *const backend,
+				    const struct cloud_event *const evt);
+
+/**
+ * @brief Cloud backend API.
+ *
+ * All cloud backends must expose this API.
+ */
+struct cloud_api {
+	int (*init)(struct cloud_backend *const backend,
+		    cloud_evt_handler_t handler);
+	int (*uninit)(const struct cloud_backend *const backend);
+	int (*connect)(const struct cloud_backend *const backend);
+	int (*disconnect)(const struct cloud_backend *const backend);
+	int (*send)(const struct cloud_backend *const backend,
+			 const struct cloud_msg *const msg);
+	int (*ping)(const struct cloud_backend *const backend);
+	int (*input)(const struct cloud_backend *const backend);
+};
+
+/**@brief Structure for cloud backend configuration. */
+struct cloud_backend_config {
+	char *name;
+	cloud_evt_handler_t handler;
+	int socket;
+};
+
+/**@brief Structure for cloud backend. */
+struct cloud_backend {
+	struct cloud_api *api;
+	struct cloud_backend_config *config;
+};
+
+/**@brief API to initialize a cloud backend. Performs all necessary
+ *	  configuration of the backend to be able to connect its online
+ *	  counterpart.
+ *
+ * param[in] cloud Pointer to cloud backend structure.
+ * param[in] ctx Pointer to backend specific data.
```

Comment:
No context parameter here, but I would say it is always good to add one, for instance as argument of the `cloud_evt_handler_t`.

Reply:
Agree, it's the plan to add it.